{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4dae569-df85-4b46-aaef-595074d8853b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ec9fcd62-e059-483c-b646-5552e56694b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/dani/miniconda3/lib/python3.9/site-packages/ete3-3.1.2-py3.7.egg/ete3/evol/parser/codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "/home/dani/miniconda3/lib/python3.9/site-packages/ete3-3.1.2-py3.7.egg/ete3/evol/parser/codemlparser.py:221: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n"
     ]
    }
   ],
   "source": [
    "import os, glob, argparse, multiprocessing, time, itertools\n",
    "from functools import partial\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from collections import Counter\n",
    "from ete3 import Tree, RectFace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2931fad2-292f-4873-b73c-0656efe2a84d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_target_taxa_groups():\n",
    "    taxa_groups_raw = [\"crassvirales\",\"NA\", \"new\", \"Viruses\", \"bact_crass_polytomy\", \"prophage\"]\n",
    "    taxa_groups = list()\n",
    "    # create the target taxa_groups to look for in the tree\n",
    "    for i in range(2, len(taxa_groups_raw)+1):\n",
    "        combinations = itertools.combinations(taxa_groups_raw, i)\n",
    "        # iterate the combinations and retain those with \"crassvirales\" in them\n",
    "        for combination in combinations:\n",
    "            if \"crassvirales\" in combination:\n",
    "                taxa_groups.append(list(combination))\n",
    "\n",
    "    return taxa_groups\n",
    "\n",
    "taxa_groups = create_target_taxa_groups()\n",
    "taxa_groups = [[\"crassvirales\"]] + taxa_groups\n",
    "\n",
    "\n",
    "def get_crassvirales_taxonomy_df(taxa_file):\n",
    "    '''\n",
    "    Parses the taxonomy and returns a dataframe, index=genomes, 3 columns (family, subfamily, genus)\n",
    "    '''\n",
    "    #\"/home/danielc/software/github/old_june/tree_functions/files/crassphages_taxonomy_terL_and_new.txt\"\n",
    "    df = pd.read_csv(taxa_file, sep=\"\\t\", index_col=0, names=[\"family\",\"subfamily\",\"genus\"])\n",
    "    df['superkingdom'] = np.where(df['family'].isin([\"new\", \"crass_env\"]), \"new\", \"crassvirales\")\n",
    "    return df\n",
    "\n",
    "def create_target_taxa_groups():\n",
    "    taxa_groups_raw = [\"crassvirales\",\"NA\", \"new\", \"Viruses\", \"bact_crass_polytomy\", \"prophage\"]\n",
    "    taxa_groups = list()\n",
    "    # create the target taxa_groups to look for in the tree\n",
    "    for i in range(2, len(taxa_groups_raw)+1):\n",
    "        combinations = itertools.combinations(taxa_groups_raw, i)\n",
    "        # iterate the combinations and retain those with \"crassvirales\" in them\n",
    "        for combination in combinations:\n",
    "            if \"crassvirales\" in combination:\n",
    "                taxa_groups.append(list(combination))\n",
    "\n",
    "    return taxa_groups\n",
    "\n",
    "def resolve_crass_bacteria_polytomies(t):\n",
    "    '''\n",
    "    Looks for clades with Crassvirales and Bacteria, all of them with 0 dist\n",
    "    '''\n",
    "    checked_leaves = list()\n",
    "    crass_polytom_leaves = t.search_nodes(dist=0.0, superkingdom=\"crassvirales\")\n",
    "    for leaf in crass_polytom_leaves:\n",
    "        if leaf.name not in checked_leaves:\n",
    "            checked_leaves.append(leaf.name)\n",
    "            # go one node up, check superkingdom of descendants\n",
    "            node_up = leaf.up\n",
    "            for node in node_up.iter_descendants():\n",
    "                if node.superkingdom in [\"Bacteria\", \"NA\"]:\n",
    "                    node.superkingdom = \"bact_crass_polytomy\"\n",
    "                    print(f\"\\t- polytomy bacteria-crass, {node.name}\")\n",
    "\n",
    "    return t\n",
    "\n",
    "def color_format_tree(t):\n",
    "    '''\n",
    "\n",
    "    '''\n",
    "\n",
    "    crassvirales_colors = {'Crevaviridae': '#EE9A00',\n",
    "                           'Intestiviridae': '#EE3B3B',\n",
    "                           'Jelitoviridae': '#006400',\n",
    "                           'Steigviridae': '#00CED1',\n",
    "                           'Suoliviridae': '#4169E1',\n",
    "                           'Tinaiviridae': '#CD2990'}\n",
    "\n",
    "    for node in t.traverse():\n",
    "        # check if it is a Crassvirales genome\n",
    "        if node.is_leaf():\n",
    "            if node.phylum in crassvirales_colors:\n",
    "                node.img_style[\"bgcolor\"] = crassvirales_colors[node.phylum]\n",
    "            else:\n",
    "                node.img_style[\"size\"] = 0\n",
    "\n",
    "            # color Bacteria as yellow\n",
    "            if node.superkingdom == \"Bacteria\":\n",
    "                node.img_style[\"bgcolor\"] = \"yellow\"\n",
    "\n",
    "    return t\n",
    "\n",
    "def read_and_annotate_tree(tree_file, ncbi_summary_df, crassvirales_taxa_df):\n",
    "    '''\n",
    "    Reads the tree of a CL with NCBI and Crassvirales sequences and assigns\n",
    "    source and taxonomy to them\n",
    "    '''\n",
    "\n",
    "    colors = {}\n",
    "\n",
    "    # read tree\n",
    "    t = Tree(tree_file, format=1)\n",
    "\n",
    "    # iterate the leaves, assi\n",
    "    for leaf in t.iter_leaves():\n",
    "        # check if it is an NCBI hit:\n",
    "        if leaf.name in ncbi_summary_df.index:\n",
    "            leaf.add_features(source=\"ncbi\",\n",
    "                              crass=str(ncbi_summary_df.loc[leaf.name, \"ncbi_crass\"]),\n",
    "                              superkingdom=ncbi_summary_df.loc[leaf.name, \"superkingdom\"],\n",
    "                              phylum=ncbi_summary_df.loc[leaf.name, \"phylum\"],\n",
    "                              #class=ncbi_summary_df.loc[leaf.name, \"class\"],\n",
    "                              order=ncbi_summary_df.loc[leaf.name, \"order\"],\n",
    "                              family=ncbi_summary_df.loc[leaf.name, \"family\"],\n",
    "                              genus=ncbi_summary_df.loc[leaf.name, \"genus\"]\n",
    "                              )\n",
    "        # if not, it is a Crassvirales sequence\n",
    "        else:\n",
    "            genome = leaf.name.split(\"|\")[0]\n",
    "            leaf.add_features(source=\"crass\",\n",
    "                              crass=str(True),\n",
    "                              superkingdom=crassvirales_taxa_df.loc[genome, \"superkingdom\"],\n",
    "                              # I put the Crassvirales family here\n",
    "                              phylum=crassvirales_taxa_df.loc[genome, \"family\"],\n",
    "                              #class=\"crass\",\n",
    "                              order=\"crass\",\n",
    "                              family=crassvirales_taxa_df.loc[genome, \"family\"],\n",
    "                              genus=crassvirales_taxa_df.loc[genome, \"genus\"]\n",
    "                              )\n",
    "\n",
    "    return t\n",
    "\n",
    "def identify_crassvirales_mrcas(t, cl_id, taxa_groups, start_cont, faces=False):\n",
    "    '''\n",
    "\n",
    "    '''\n",
    "\n",
    "    monophyletic_clades_leaves = dict()\n",
    "    monophyletic_clades = dict()\n",
    "    mrcas_ids_to_clean = list()\n",
    "    cont = 0\n",
    "    for taxa_group in taxa_groups:\n",
    "        mrcas = t.get_monophyletic(target_attr=\"superkingdom\", values=taxa_group)\n",
    "        if mrcas:\n",
    "            for mrca in mrcas:\n",
    "                cont += 1\n",
    "                mrca_id = f\"mrca_{cont}\"\n",
    "                # get names of the leaves\n",
    "                mrca_leaves_names = {leaf.name for leaf in mrca.iter_leaves()}\n",
    "                if monophyletic_clades_leaves:\n",
    "                    for target_mrca, target_leaves_names in monophyletic_clades_leaves.items():\n",
    "                        tmp = target_leaves_names.union(mrca_leaves_names)\n",
    "                        if len(tmp) == len(mrca_leaves_names):\n",
    "                            mrcas_ids_to_clean.append(target_mrca)\n",
    "\n",
    "                monophyletic_clades_leaves[mrca_id] = mrca_leaves_names\n",
    "                monophyletic_clades[mrca_id] = mrca\n",
    "\n",
    "\n",
    "    # remove MRCAs already contained in higher MRCAs\n",
    "    mrcas_ids_to_clean = list(set(mrcas_ids_to_clean))\n",
    "    for mrca_id in mrcas_ids_to_clean:\n",
    "        del monophyletic_clades[mrca_id]\n",
    "        del monophyletic_clades_leaves[mrca_id]\n",
    "\n",
    "\n",
    "    # change mrca_ids to lower numbers\n",
    "    final_monophyletic_clades = dict()\n",
    "    final_cont = start_cont\n",
    "    for mrca_id, mrca in monophyletic_clades.items():\n",
    "        final_monophyletic_clades[f\"mrca_{final_cont}\"] = mrca\n",
    "        if faces:\n",
    "            mrca.add_features(mrca_id=f\"mrca_{final_cont}\")\n",
    "            face = RectFace(5,5, \"lime\",\"lime\", label=f\"mrca_{final_cont}\")\n",
    "            mrca.add_face(face, 0, \"float\")\n",
    "        final_cont +=1\n",
    "\n",
    "    #if faces:\n",
    "    #    pdf_outfile = f\"/home/danielc/projects/Bas_phages/5_nr_screening/4_merged_ncbi_crassvirales/4_MRCAs_taxa/{cl_id}.pdf\"\n",
    "    #    t.render(pdf_outfile)\n",
    "\n",
    "    return t, final_monophyletic_clades, final_cont\n",
    "\n",
    "def identify_outgroup_seqs(t, mrcas):\n",
    "    '''\n",
    "    '''\n",
    "    outgroup_seqs = list()\n",
    "    mrcas_outgroups = dict()\n",
    "    \n",
    "    # get the most distant Bacteria node to the mrca\n",
    "    for mrca_id, mrca in mrcas.items():\n",
    "        bacteria_leaves_dists = {bacteria_leaf:mrca.get_distance(bacteria_leaf) for bacteria_leaf in t.search_nodes(superkingdom=\"Bacteria\")}\n",
    "        # get the most distant NA node to the mrca\n",
    "        NA_leaves_dists = {bacteria_leaf:mrca.get_distance(bacteria_leaf) for bacteria_leaf in t.search_nodes(superkingdom=\"NA\")}\n",
    "\n",
    "        # merge both dictionaries\n",
    "        bacteria_leaves_dists.update(NA_leaves_dists)\n",
    "\n",
    "        if bacteria_leaves_dists:\n",
    "            # sort by distance\n",
    "            bacteria_leaves_dists = {node: distance for node, distance in sorted(bacteria_leaves_dists.items(), key=lambda item: item[1], reverse=True)}\n",
    "            farthest_bact_name = list(bacteria_leaves_dists.keys())[0].name\n",
    "            outgroup_seqs.append(farthest_bact_name)\n",
    "            \n",
    "            mrcas_outgroups[mrca_id] = farthest_bact_name\n",
    "\n",
    "    outgroup_seqs = list(set(outgroup_seqs))\n",
    "\n",
    "    return outgroup_seqs, mrcas_outgroups\n",
    "\n",
    "def get_rooted_trees(outgroup_seqs, tree_file, ncbi_summary_df, crassvirales_taxa_df):\n",
    "    '''\n",
    "    '''\n",
    "    rooted_trees = dict()\n",
    "\n",
    "    for outgroup_seq in outgroup_seqs:\n",
    "        # read the CL tree\n",
    "        t = read_and_annotate_tree(tree_file, ncbi_summary_df, crassvirales_taxa_df)\n",
    "        # add colors\n",
    "        t = color_format_tree(t)\n",
    "        # resolve Crassvirales-Bacteria polytomies\n",
    "        t = resolve_crass_bacteria_polytomies(t)\n",
    "\n",
    "        outgroup_node = t.search_nodes(name=outgroup_seq)[0]\n",
    "        t.set_outgroup(outgroup_node)\n",
    "        rooted_trees[outgroup_seq] = t\n",
    "\n",
    "    return rooted_trees\n",
    "\n",
    "\n",
    "def get_taxa_sister(sister_clade, polytomy_nodes):\n",
    "    '''\n",
    "    Account also for the Bacteria politomies in the mrca\n",
    "    '''\n",
    "    superkingdoms = [leaf.superkingdom for leaf in sister_clade.iter_leaves()]\n",
    "    if polytomy_nodes:\n",
    "        superkingdoms += [\"Bacteria\" for polytomy in polytomy_nodes]\n",
    "\n",
    "\n",
    "    # get the percentage of each phylum\n",
    "    counter = Counter(superkingdoms)\n",
    "    total = sum(counter.values())\n",
    "    # store each phylum along with its percentage and count\n",
    "    superkingdoms_percent = [(superkingdom, round(count/total, 3)) for superkingdom, count in counter.items()]\n",
    "    superkingdoms_percent = sorted(superkingdoms_percent, key=lambda superkingdom: superkingdom[1], reverse=True )\n",
    "    superkingdoms_percent = \";\".join([f\"{superkingdom[0]}:{superkingdom[1]}\" for superkingdom in superkingdoms_percent])\n",
    "\n",
    "\n",
    "    # store the phyla for the Bacteria hits\n",
    "    phyla = [leaf.phylum for leaf in sister_clade.iter_leaves() if leaf.superkingdom == \"Bacteria\"]\n",
    "    if polytomy_nodes:\n",
    "        phyla += [polytomy.phylum for polytomy in polytomy_nodes]\n",
    "\n",
    "    counter = Counter(phyla)\n",
    "    total = sum(counter.values())\n",
    "    # store each phylum along with its percentage and count\n",
    "    phyla_percent = [(phylum, round(count/total, 3)) for phylum, count in counter.items()]\n",
    "    phyla_percent = sorted(phyla_percent, key=lambda phylum: phylum[1], reverse=True )\n",
    "    phyla_percent = \";\".join([f\"{phylum[0]}:{phylum[1]}\" for phylum in phyla_percent])\n",
    "\n",
    "\n",
    "    # store order for the Bacteria hits\n",
    "    orders = [f\"{leaf.phylum}_{leaf.order}\" for leaf in sister_clade.iter_leaves() if leaf.superkingdom == \"Bacteria\"]\n",
    "    if polytomy_nodes:\n",
    "        orders += [f\"{polytomy.phylum}_{polytomy.order}\" for polytomy in polytomy_nodes]\n",
    "\n",
    "    counter = Counter(orders)\n",
    "    total = sum(counter.values())\n",
    "    # store each phylum along with its percentage and count\n",
    "    orders_percent = [(order, round(count/total, 3)) for order, count in counter.items()]\n",
    "    orders_percent = sorted(orders_percent, key=lambda order: order[1], reverse=True )\n",
    "    orders_percent = \";\".join([f\"{order[0]}:{order[1]}\" for order in orders_percent])\n",
    "\n",
    "    return superkingdoms_percent, phyla_percent, orders_percent\n",
    "\n",
    "def process_MRCA(mrca):\n",
    "    '''\n",
    "    '''\n",
    "    polytomy_nodes = [leaf for leaf in mrca.iter_leaves() if leaf.superkingdom == \"bact_crass_polytomy\"]\n",
    "    polytomy_seqs = \"\"\n",
    "    if polytomy_nodes:\n",
    "        polytomy_seqs = \",\".join([leaf.name for leaf in mrca.iter_leaves() if leaf.superkingdom == \"bact_crass_polytomy\"])\n",
    "\n",
    "    seqs_mrca = \",\".join(mrca.get_leaf_names())\n",
    "    n_seqs_mrca = len(mrca.get_leaf_names())\n",
    "\n",
    "    sister = mrca.get_sisters()\n",
    "    if len(sister) > 1:\n",
    "        print(\"wuh?\")\n",
    "    seqs_sister = \",\".join(sister[0].get_leaf_names())\n",
    "    n_seqs_sister = len(sister[0].get_leaf_names())\n",
    "    n_seqs_bacteria = len([leaf.name for leaf in sister[0].iter_leaves() if leaf.superkingdom == \"Bacteria\"])\n",
    "    bacteria_seqs_sister = \",\".join([leaf.name for leaf in sister[0].iter_leaves() if leaf.superkingdom == \"Bacteria\"])\n",
    "    sister_support = mrca.up.name\n",
    "\n",
    "    superkingdoms_sister, bact_phyla_sister, bact_order_sister = get_taxa_sister(sister[0], polytomy_nodes)\n",
    "\n",
    "    return polytomy_seqs, n_seqs_mrca, seqs_mrca, sister_support, n_seqs_sister, seqs_sister, n_seqs_bacteria, bacteria_seqs_sister, superkingdoms_sister, bact_phyla_sister, bact_order_sister\n",
    "\n",
    "def process_tree(summary_dir, crassvirales_taxa_df, taxa_groups, tree_file):\n",
    "\n",
    "    # get CL_id\n",
    "    cl_id = os.path.basename(tree_file).split(\"_ncbi\")[0]\n",
    "\n",
    "    # read summary file of the CL to know wheter it needs to be processed for Bacteria\n",
    "    summary_file = f\"{summary_dir}/{cl_id}.summary\"\n",
    "    ncbi_summary_df = pd.read_csv(summary_file, sep=\"\\t\", header=0, index_col=0, low_memory=False)\n",
    "    ncbi_summary_df = ncbi_summary_df[ncbi_summary_df.included]\n",
    "    # check if there are Bacteria hits\n",
    "    if \"Bacteria\" in ncbi_summary_df.superkingdom.to_list():\n",
    "        # replace nan by \"NA\"\n",
    "        ncbi_summary_df = ncbi_summary_df.fillna(\"NA\")\n",
    "\n",
    "        # read the CL tree\n",
    "        t = read_and_annotate_tree(tree_file, ncbi_summary_df, crassvirales_taxa_df)\n",
    "        # add colors\n",
    "        t = color_format_tree(t)\n",
    "        # resolve Crassvirales-Bacteria polytomies\n",
    "        t = resolve_crass_bacteria_polytomies(t)\n",
    "        # identify Crassvirales MRCAs in the tree\n",
    "        start_cont = 1\n",
    "        t, mrcas, cont = identify_crassvirales_mrcas(t, cl_id, taxa_groups, start_cont, faces=True)\n",
    "\n",
    "        # identify all the distant Bacteria/NA sequences and get the trees rooted with them\n",
    "        outgroup_seqs, mrcas_outgroups = identify_outgroup_seqs(t, mrcas)\n",
    "        rooted_trees = get_rooted_trees(outgroup_seqs, tree_file, ncbi_summary_df, crassvirales_taxa_df)\n",
    "        print(mrcas_outgroups)\n",
    "        \n",
    "        t.show()\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "513b481e-f144-41cf-88a0-bfc90af79550",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "205fa979-158e-421e-8bec-30e61f0fbd59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get crassvirales taxonomy\n",
    "crassvirales_taxa_file = \"/home/dani/encode_home//projects/Bas_phages/crass_protein_families_scripts/tree_functions/crassphages_taxonomy_terL_and_new.txt\"\n",
    "crassvirales_taxa_df = get_crassvirales_taxonomy_df(crassvirales_taxa_file)\n",
    "\n",
    "# target taxa clades to look for in the trees\n",
    "taxa_groups = create_target_taxa_groups()\n",
    "taxa_groups = [[\"crassvirales\"]] + taxa_groups # add crassvirales at the beginning\n",
    "\n",
    "# list trees\n",
    "trees_dir = \"/home/dani/encode_home/projects/Bas_phages/5_nr_screening/4_merged_ncbi_crassvirales/2_trees/\"\n",
    "tree_files = glob.glob(f\"{trees_dir}/cl_019*.nw\")\n",
    "\n",
    "summary_dir = \"/home/dani/encode_home/projects/Bas_phages/5_nr_screening/2_hits_summary/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3da8bf06-b8bf-4653-a1bf-0210b10e4984",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t- polytomy bacteria-crass, WP_203042047.1\n",
      "\t- polytomy bacteria-crass, QWM89166.1\n",
      "\t- polytomy bacteria-crass, WP_203042047.1\n",
      "\t- polytomy bacteria-crass, QWM89166.1\n",
      "\t- polytomy bacteria-crass, WP_203042047.1\n",
      "\t- polytomy bacteria-crass, QWM89166.1\n",
      "{'mrca_1': 'MBQ7819935.1', 'mrca_2': 'MBQ7819935.1', 'mrca_3': 'MBQ7819935.1', 'mrca_4': 'MBQ7819935.1', 'mrca_5': 'MBQ7819935.1', 'mrca_6': 'WP_194270469.1', 'mrca_7': 'WP_194270469.1', 'mrca_8': 'WP_194270469.1', 'mrca_9': 'MBQ7819935.1', 'mrca_10': 'MBQ7819935.1', 'mrca_11': 'MBQ7819935.1', 'mrca_12': 'WP_194270469.1', 'mrca_13': 'WP_194270469.1', 'mrca_14': 'MBQ7819935.1', 'mrca_15': 'MBQ7819935.1', 'mrca_16': 'MBQ7819935.1', 'mrca_17': 'WP_194270469.1'}\n"
     ]
    }
   ],
   "source": [
    "for tree_file in tree_files:\n",
    "    if \"cl_0193\" in tree_file:\n",
    "        process_tree(summary_dir, crassvirales_taxa_df, taxa_groups, tree_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92056133-71f5-4696-997a-f86bcdcd67b0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c4c10d1-b597-40a5-b6a2-566aefd24ac2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64553eff-5c7e-40f4-8e30-d8a6c63620fc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0711966f-1772-4526-8ef6-3f376a99b042",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bc8b9e2-72ee-40f3-b136-9f0ff09c5899",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3f1abc8-c8c3-4f48-82f6-dffe16350632",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f3847c4-c569-4262-a574-2c9867bf849e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1252967-06d6-4d45-a5a4-94df540a3d55",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64b7e1bf-e4d7-48fd-be4f-6305569f8306",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "300e2a15-7c61-4355-96c3-2405be9cd4b1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c13e88e-99c2-4649-b3b0-7faaf47096de",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
